keps/0012-label-index.md

```diff
@@ -0,0 +1,51 @@
+---
+kep-number: 12
+title: Index labels to support efficient queries by label selectors
+authors:
+  - "@jian-he, @ccding"
+owning-sig: sig-api-machinery
+reviewers:
+  - TBD
+approvers:
+  - TBD
+creation-date: "2018-05-31"
+last-updated: "2018-05-31"
+status: provisional
+---
+
+## Summary
+
+This proposal presents solutions to make the queries by label selectors in the
+list API efficient.
+
+## Motivation
+
+Currently, only namespace index is supported in the cache. Controllers such as
+ReplicaSet, when selecting pods, need to select all the pods by namespace and
+then select the ones wanted filtering by index. While running a large number
+of pods, we observed that listing all pods and then filtering takes several
+seconds. This proposal presents solutions to improve this.
+
+
+### Goals
+
+Make the queries by label selectors in the list API efficient.
+
+### Proposal
+
+1. Adds label based index in controllers to facilitate lookup by label.
```

Comment:
This is not really a design :)

How will this be accomplished? What components will change? How will you support all of our selector options, in particular IN?

(I'm not saying these questions are not answerable and we can't do it, I'm saying they are and this document should provide the answers)

Reply:
Oh I missed the word "controllers", you can already add indexes. Which controllers do you want to change?

Reply:
I am not sure if you need to have a super detailed design in the kep itself but it needs to be somewhere.